### ISSUE TITLE
Some cue errors are not caught before running

### ISSUE THREAD

**Issue by shykes** [dagger/dagger#493]
One big selling point of Cue is that you can catch many errors in your configuration *before* applying it. However, certain errors which *should* be caught before running (or at least feel like they should be caught), are not.

So far I have identified one category of errors which is *not* caught by Cue before execution:
* References to a non-existing definition in an imported package

For example, I expect this snippet to fail before execution, but it does not:

```
import (
   “strings” // Package name does not matter. Builtin or external package doesn’t matter.
)

// #Nope field does not exist in the package
foo: strings.#Nope
```

In summary, references to non-existing fields in another package are *not* considered an error at build time, but will in fact resolve to an error when queried at run time.

## Related
- [ ] https://github.com/dagger/dagger/issues/2331

**Comment by helderco** [comment]
> So far I have identified one category of errors which is *not* caught by Cue before execution:
> * References to a non-existing definition in an imported package

@shykes, here's more: "references to incomplete errors", which more broadly should include the above.

Example (#1885):
```cue
dagger.#Plan & {
	actions: {
		test: core.#Nop & {
			input: actions.nonexistent
		}
	}
}
```
Or this (#1857):

```cue
dagger.#Plan & {
	actions: {
		write: actions.test.target.output
		test: {
			dep: core.#Nop & {
				input: actions.nonexistent
			}
			target: core.#Nop & {
				input: dep.output
			}
		}
	}
}
```
Or this (#1781):
```cue
client: filesystem: ...: write: contents: "\(actions.valid.path)" // This doesn't work
```
These run silently and it's tripping users because of typos.

**Comment by helderco** [comment]
Here's a compilation of these issues (perhaps more):

- https://github.com/dagger/dagger/issues/1635
- https://github.com/dagger/dagger/issues/1781
- https://github.com/dagger/dagger/issues/1857
- https://github.com/dagger/dagger/issues/1885
- https://github.com/dagger/dagger/issues/1990

Let's close them as duplicates?

**Comment by shykes** [comment]
I'm fine with closing duplicates as long as they are actually duplicates (ie. solving one is certain to solve the other). If you feel up for triaging them, please do & thank you! Those errors are a pain, anything that gets us closer to solving them is appreciated.

**Comment by helderco** [comment]
They are various **side-effects** of the same underlying issue which is referencing "incomplete errors" not failing early (which is what this issue is about).

I'll keep each particular side-effect open so we can remember to test them later (avoid regressions), close their duplicates and track all here at a high level.

**Comment by shykes** [comment]
I just want to ping this again. This is a horrible bug and easily in my top 3 of most painful DX issues. If upstream can't or won't fix it, we must find some way to limit the damage downstream.

**Comment by helderco** [comment]
> @helderco Can you help me out filling the integration tests with all reported cases of this issue, verifying it actually fixes all of them?

Sure 🙂

**Comment by myitcv** [comment]
I've commented over in https://github.com/cue-lang/cue/issues/629#issuecomment-1111822832. Please can you help to establish whether the issues Dagger are seeing are covered by the cases in the original description of https://github.com/cue-lang/cue/issues/629 or not? That will then help to confirm whether the proposal in https://github.com/cue-lang/cue/issues/629#issuecomment-1111867269 is sufficient or not.

**Comment by helderco** [comment]
> @helderco Can you help me out filling the integration tests with all reported cases of this issue, verifying it actually fixes all of them?

There's only one remaining issue from my list above: #1885. I confirm that's covered in #2334.

**Comment by mircubed** [comment]
based on this comment from @shykes , @aluzzardi and @helderco please check if this needs to be re-opened? 

discord comment -https://discord.com/channels/707636530424053791/796905486145683506/989285536420216872
